src/Microsoft.SqlTools.ServiceLayer/LanguageServices/ConnectedBindingContext.cs

```diff
@@ -162,68 +180,68 @@ public ParseOptions ParseOptions
         {
-            if (engineType == DatabaseEngineType.SqlAzureDatabase)
+            if (server.DatabaseEngineType == DatabaseEngineType.SqlAzureDatabase)
             {
                 return DatabaseCompatibilityLevel.Azure;
             }
-            int versionMajor = Math.Max(serverVersion.Major, 8);
 
-            switch (versionMajor)
+            // Get the actual compat level of the database we're connected to
+            switch (server.Databases[server.ConnectionContext.DatabaseName].CompatibilityLevel)
             {
-                case 8:
+                case SMO.CompatibilityLevel.Version80:
                     return DatabaseCompatibilityLevel.Version80;
-                case 9:
+                case SMO.CompatibilityLevel.Version90:
                     return DatabaseCompatibilityLevel.Version90;
-                case 10:
+                case SMO.CompatibilityLevel.Version100:
                     return DatabaseCompatibilityLevel.Version100;
-                case 11:
+                case SMO.CompatibilityLevel.Version110:
                     return DatabaseCompatibilityLevel.Version110;
-                case 12:
+                case SMO.CompatibilityLevel.Version120:
                     return DatabaseCompatibilityLevel.Version120;
-                case 13:
+                case SMO.CompatibilityLevel.Version130:
                     return DatabaseCompatibilityLevel.Version130;
-                case 14:
+                case SMO.CompatibilityLevel.Version140:
                     return DatabaseCompatibilityLevel.Version140;
-                case 15:
+                case SMO.CompatibilityLevel.Version150:
                     return DatabaseCompatibilityLevel.Version150;
                 default:
                     return DatabaseCompatibilityLevel.Current;
             }
         }
 
         /// <summary>
-        /// Gets the transaction sql version from a server version
+        /// Gets the transaction sql version for a given server connection
         /// </summary>
-        /// <param name="serverVersion"></param>
-        private static TransactSqlVersion GetTransactSqlVersion(DatabaseEngineType engineType, ServerVersion serverVersion)
+        /// <param name="server"></param>
+        private static TransactSqlVersion GetTransactSqlVersion(SMO.Server server)
         {
-            if (engineType == DatabaseEngineType.SqlAzureDatabase)
+            if (server.DatabaseEngineType == DatabaseEngineType.SqlAzureDatabase)
             {
                 return TransactSqlVersion.Azure;
             }
 
-            int versionMajor = Math.Max(serverVersion.Major, 9);
-
-            switch (versionMajor)
+            // Use the compat level of the database we're connected to for determing what verison
+            // of T-SQL to support
+            switch (server.Databases[server.ConnectionContext.DatabaseName].CompatibilityLevel)
```

Comment:
the compat level of the database doesn't limit the DDL syntax. We should use the Math.Max equivalent of (serverVersion.Major, CompatibilityLevel) because there are now SQL editions whose DDL version is higher than their engine version. Such a check will miss an MI instance set to a lower compat level but the alternative is to start checking for specific editions and we don't know a priori when a given edition will start supporting new DDL versions. Maybe we could check the compat level of "master" specifically and fall back to the user database compat level only if the user can't connect to master. I think master will always have the highest compat level for a given edition.